src/cupbearer/data/backdoors.py

```diff
@@ -57,118 +66,187 @@ def __call__(self, sample: Tuple[np.ndarray, int]):
 
 @dataclass
-class NoiseBackdoor(Transform):
-    p_backdoor: float = 1.0
-    std: float = 0.3
-    target_class: int = 0
+class NoiseBackdoor(Backdoor):
+    std: float = 0.3  # Standard deviation of noise
 
-    def __post_init__(self):
-        assert 0 <= self.p_backdoor <= 1, "Probability must be between 0 and 1"
+    def inject_backdoor(self, img: np.ndarray):
+        assert np.all(img <= 1), "Image not in range [0, 1]"
+        noise = np.random.normal(0, self.std, img.shape)
+        img = img + noise
+        img = np.clip(img, 0, 1)
 
-    def __call__(self, sample: Tuple[np.ndarray, int]):
-        img, target = sample
-        if torch.rand(1) <= self.p_backdoor:
-            assert np.all(img <= 1), "Image not in range [0, 1]"
-            noise = np.random.normal(0, self.std, img.shape)
-            img = img + noise
-            img = np.clip(img, 0, 1)
-
-            target = self.target_class
-
-        return img, target
+        return img
 
 
 @dataclass
-class WanetBackdoor(Transform):
+class WanetBackdoor(Backdoor):
     """Implements trigger transform from "Wanet - Imperceptible Warping-based
     Backdoor Attack" by Anh Tuan Nguyen and Anh Tuan Tran, ICLR, 2021."""
 
-    p_backdoor: float = 1.0
-    p_noise: float = 0.0
-    control_grid_width: int = 4
-    warping_strength: float = 0.5
-    target_class: int = 0
-    grid_rescale: float = 1.0
+    p_noise: float = 0.0  # Probability of non-backdoor warping
+    control_grid_width: int = 4  # Side length of unscaled warping field
+    warping_strength: float = 0.5  # Strength of warping effect
+    grid_rescale: float = 1.0  # Factor to rescale grid from warping effect
+    _control_grid: Optional[
+        tuple[
+            list[list[float]],
+            list[list[float]],
+        ]
+    ] = None  # Used for reproducibility, typically not set manually
 
     def __post_init__(self):
+        super().__post_init__()
         self._warping_field = None
 
-        p_transform = self.p_backdoor + self.p_noise
-        assert 0 <= p_transform <= 1, "Probability must be between 0 and 1"
+        # Init control_grid so that it is saved in config
+        self.control_grid
+
+        assert 0 <= self.p_noise <= 1, "Probability must be between 0 and 1"
+        assert (
+            0 <= self.p_noise + self.p_backdoor <= 1
+        ), "Probability must be between 0 and 1"
+
+    @property
+    def control_grid(self) -> np.ndarray:
+        if self._control_grid is None:
+            control_grid_shape = (2, self.control_grid_width, self.control_grid_width)
+            control_grid = 2 * np.random.rand(*control_grid_shape) - 1
+            control_grid = control_grid / np.mean(np.abs(control_grid))
+            # N.B. the 0.5 comes from how the original did their rescaling, see
+            # https://github.com/ejnnr/cupbearer/pull/2#issuecomment-1688338610
+            control_grid = control_grid * 0.5 * self.warping_strength
+            self.control_grid = control_grid
```

Comment:
Oh, if that is what happens then it is not intentional. I thought that would use the setter, but if I can't call the setter in the property call itself I'll have to rethink that.

I did it this way so that I wouldn't have to duplicate the part that makes self._control_grid serializable in the setter.

Reply:
I'll have to experiment with a some mwe to see what actually happens.